### PR TITLE
Code Review: Reported crash on macOS Sierra when selecting multiple objects (#9758)

### PR DESCRIPTION
**Pull request by pieteromvlee**
Code review for Reported crash on macOS Sierra when selecting multiple objects (#9758):

> **Link to HockeyApp**
> 
> https://rink.hockeyapp.net/manage/apps/71946/app_versions/224/crash_reasons/126318865?type=crashes
> 
> **Stacktrace**
> 
> ````
> 1	AppKit	-[NSApplication _crashOnException:] + 109
> 2	AppKit	-[NSApplication _showException:] + 747
> 3	AppKit	-[NSApplication reportException:] + 359
> 4	AppKit	-[NSApplication run] + 1148
> 5	AppKit	NSApplicationMain + 1237
> 6	libdyld.dylib	start + 1
> ````
> 
> **Description**
> 
> >Crash after select few objects with shift
> 
> ````
> Exception Name: NSRangeException
> Description: *** -[NSArray subarrayWithRange:]: range {2, 3} extends beyond bounds [0 .. 2]
> User Info: (null)
> 
> 0   CoreFoundation                      0x00007fff9dcf874b __exceptionPreprocess + 171
> 1   libobjc.A.dylib                     0x00007fffb152a80a objc_exception_throw + 48
> 2   CoreFoundation                      0x00007fff9dc92638 -[NSArray subarrayWithRange:] + 616
> 3   Sketch Beta                         0x000000010abfc100 Sketch Beta + 1200384
> 4   Sketch Beta                         0x000000010aadd1bb Sketch Beta + 25019
> 5   Sketch Beta                         0x000000010aca3f02 Sketch Beta + 1888002
> 6   AppKit                              0x00007fff9bcc7ea9 -[NSToolbarItem _validateAsCommonItem:] + 183
> 7   AppKit                              0x00007fff9ba59777 -[NSToolbarItem validate] + 115
> 8   AppKit                              0x00007fff9bad4bf0 -[NSToolbar validateVisibleItems] + 212
> 9   AppKit                              0x00007fff9bb3644f -[NSToolbar _autovalidateVisibleToolbarItems] + 87
> 10  AppKit                              0x00007fff9bb363c1 __53-[NSToolbarView _scheduleDelayedValidationAfterTime:]_block_invoke + 156
> 11  CoreFoundation                      0x00007fff9dc79454 __CFRUNLOOP_IS_CALLING_OUT_TO_A_TIMER_CALLBACK_FUNCTION__ + 20
> 12  CoreFoundation                      0x00007fff9dc790e3 __CFRunLoopDoTimer + 1075
> 13  CoreFoundation                      0x00007fff9dc78c3a __CFRunLoopDoTimers + 298
> 14  CoreFoundation                      0x00007fff9dc705f1 __CFRunLoopRun + 2081
> 15  CoreFoundation                      0x00007fff9dc6fb6d CFRunLoopRunSpecific + 285
> 16  HIToolbox                           0x00007fff9d24227c RunCurrentEventLoopInMode + 240
> 17  HIToolbox                           0x00007fff9d2420b1 ReceiveNextEventCommon + 432
> 18  HIToolbox                           0x00007fff9d241ee6 _BlockUntilNextEventMatchingListInModeWithFilter + 71
> 19  AppKit                              0x00007fff9b9b1101 _DPSNextEvent + 1093
> 20  AppKit                              0x00007fff9c0b6a0c -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 670
> 21  AppKit                              0x00007fff9b9a5a3e -[NSApplication run] + 929
> 22  AppKit                              0x00007fff9b971367 NSApplicationMain + 1237
> 23  libdyld.dylib                       0x00007fffb1e03285 start + 1
> 24  ???                                 0x0000000000000001 0x0 + 1
> ````
> 
> **Steps**
> 
> 1. Open the [example document](https://www.dropbox.com/s/jctp5p8ocelnfx5/Sketch%203.9%20b2%20-%20Selection%20Issue.sketch.zip?dl=0)
> 2. Select the yellow rectangle in the left-hand group
> 3. Still selecting the first yellow rectangle, hold CMD+Shift then select the yellow rectangle in the right-hand group
> 
> 
> **Notes**
> 
> * Example document and screen recording demo in Dropbox folder

Connect to BohemianCoding/Sketch#9758.

### CODE REVIEW

File: Source/NSArray+BCFoundation.m [hunk 1]
```diff
@@ -254,7 +254,7 @@ - (NSUInteger)indexOfSubArray:(NSArray *)array {
   
   if ([self count] < [array count])
     return NSNotFound;
-  else if (index != NSNotFound && [[self subarrayWithRange:NSMakeRange(index, [array count])] isEqual:array])
```
Comment (randomsequence) [inline Source/NSArray+BCFoundation.m]: This method is a bit of a tangle! There are multiple method calls on [self count] and [array count] and an undesirable early return, and an unused method return value when that early return is taken. How about:

```objc
- (NSUInteger)indexOfSubArray:(NSArray *)array {
  id first = [array firstObject];
  NSUInteger index = [self indexOfObject:first];

  NSUInteger selfCount = [self count];  
  NSUInteger arrayCount = [array count];
  
  if (selfCount > arrayCount) {
    index = [self indexOfObject:first];
    if (index == NSNotFound 
        || index+arrayCount > selfCount 
        || ![[self subarrayWithRange:NSMakeRange(index, arrayCount)] isEqual:array]) {
        index = NSNotFound;
        }
  }

    return index;
}
```

**Comment by randomsequence** [comment]
Yay for tests! Reviewed with a suggestion.

**Comment by pieteromvlee** [comment]
I have no problem with multiple returns, but I did realise what I effectively did was for checking if one range intersected another range and we have a function for that. So I simplified the method quite a bit :)

**Comment by randomsequence** [comment]
Nice improvement 👍

**Pull request merged**